Add SQL Server access and GUI guides (#10293)
gravitational/teleport: docs/pages/database-access/guides/sql-server-ad.mdx

---
title: Database Access with Microsoft SQL Server with Active Directory authentication (Preview)
description: How to configure Teleport Database Access with Microsoft SQL Server with Active Directory authentication.
---

<Details
  title="Version warning"
  opened={true}
  scope={["oss", "enterprise"]}
  scopeOnly={true}
  min="9.0"
>
  Database access for Microsoft SQL Server with Active Directory authentication
  is available starting from Teleport `9.0`.
</Details>

<Admonition type="warning" title="Preview">
  Database access for Microsoft SQL Server is currently in a Preview mode and
  does not include audit logging of database query activity.
</Admonition>

This guide will help you to:

- Install and configure Teleport.

> - Set up access to your SQL Server using Active Directory authentication.

Reviewer: ```suggestion
- Set up access to SQL Server using Active Directory authentication.
```
or "your SQL Server deployment"?

> - Connect to the database server through Teleport.

Reviewer: Is "the database server" the same as "SQL Server"? Or is this referring to the Teleport Database Service?

This guide will focus on Amazon RDS for SQL Server using AWS-managed Active
Directory authentication.

## Prerequisites

- Teleport version >= `9.0`.
- A SQL Server database with Active Directory authentication enabled.
- A Windows machine joined to the same Active Directory domain as the database.

> - A Linux node joined to the same Active Directory domain as the database. This

Reviewer: Should we capitalize "node" here? I.e., is this a Teleport Node or a Linux VM more generally?

>   guide will walk you through the joining steps if you don't have one.

Reviewer: Does "one" refer to "A Linux node"?

Author: Yes.

> ## Step 1/7. Set up Teleport Auth and Proxy

Reviewer: Does this guide support Teleport Cloud? If so, we may want to remove this section and list "A running Teleport Auth and Proxy Service, whether in the Cloud or self hosted" (or similar) in the prerequisites.

Author: It would be a bit inconsistent with other guides that try and explain how to install Teleport as well (and avoid having to send users around between different guides). I've added an admonition explaining that this step is not needed for Cloud users - would that work?

Reviewer: Now that we've merged [this commit](https://github.com/gravitational/teleport/commit/0499c6ef8e33d45b536521983198e3476a4d13c2#diff-b95c20e85ebd3429ca8c4a2f99ec113280dd50b784f4f3ff972f4c3f39a304a3), I think it's okay to remove this Admonition and keep this section, since it includes a tab for Cloud users.

(!docs/pages/includes/database-access/start-auth-proxy.mdx!)

## Step 2/7. Create Teleport user

(!docs/pages/includes/database-access/create-user.mdx!)

## Step 3/7. Join Linux node to Active Directory

<Admonition type="note">

>   You can skip this step if you already have a Linux node joined to the same

Reviewer: Same question re: "node" as earlier (also applies to other mentions of "node" in this guide)

Author: Replied above.

  Active Directory domain as your SQL Server.
</Admonition>

The Linux node where the Database Service will run must be joined to the same
Active Directory domain as the SQL Server database.

Note that in order to be able to join, the Linux node must be able to resolve
your Active Directory fully-qualified domain name. For example, for AWS-managed

> AD use nameservers provided under networking details on the directory's overview

Reviewer: Also, is "Networking Details" the title of something in the UI?

    $ sudo apt-get update
    $ sudo ACCEPT_EULA=Y apt-get install -y adutil
    ```
  </TabItem>
  <TabItem label="Ubuntu 20.04">
    ```bash
    $ sudo wget -qO /etc/apt/trusted.gpg.d/microsoft.asc https://packages.microsoft.com/keys/microsoft.asc
    $ sudo curl https://packages.microsoft.com/config/ubuntu/20.04/prod.list | sudo tee /etc/apt/sources.list.d/msprod.list
    $ sudo apt-get update
    $ sudo ACCEPT_EULA=Y apt-get install -y adutil
    ```
  </TabItem>
  <TabItem label="RHEL / CentOS 7">
    ```bash
    $ sudo curl -o /etc/yum.repos.d/msprod.repo https://packages.microsoft.com/config/rhel/8/prod.repo
    $ sudo ACCEPT_EULA=Y yum install -y adutil
    ```
  </TabItem>
</Tabs>

Log into your Active Directory using the `kinit` command:

```bash
$ kinit admin@EXAMPLE.COM
```

Use the `adutil keytab create` command to generate keytab entries for each
Active Directory user that will be connecting to the SQL Server database:

```bash
$ adutil keytab create teleport.keytab alice
$ adutil keytab create teleport.keytab bob
```

You will be prompted to enter each user's password. All keytab entries will
be merged into the same `teleport.keytab` file.

> <Admonition type="note" title="SPN">

Reviewer: I think we should use a more descriptive title so users don't skip over this box.

    --ad-realm=EXAMPLE.COM \
    --ad-spn=MSSQLSvc/sqlserver.example.com:1433 \
    --labels=env=dev
  ```

Provide Active Directory parameters:

| Flag | Description |
| ---- | ----------- |
| `--ad-keytab` | Path to Kerberos keytab file generated above. |
| `--ad-realm` | Kerberos realm (Active Directory domain) that SQL Server is joined. |
| `--ad-spn` | Service Principal Name for SQL Server to fetch Kerberos tickets for. |

<Admonition type="tip">
  You can start Teleport Database Service using a configuration file instead of
  CLI flags. See [YAML reference](../reference/configuration.mdx).
</Admonition>

### Service Principal Name

You can use `adutil spn search` command to see the SPNs registered for your SQL
Server. Typically, they take a form of `MSSQLSvc/<name>.<ad-domain>:<port>`.

For example, AWS RDS SQL Server named `sqlserver` and joined to an AWS managed
Active Directory domain `EXAMPLE.COM` will have the following SPNs registered:

```bash
$ adutil spn search
...
CN=EC2AMAZ-4KN05DU,OU=RDS,OU=AWS Reserved,DC=example,DC=com:
    MSSQLSvc/sqlserver.example.com:1433
    MSSQLSvc/EC2AMAZ-4KN05DU.example.com:1433
    MSSQLSvc/EC2AMAZ-4KN05DU.example.com
...
```

> Alternatively, you can look SPNs up in the Attribute Editor of Active Directory

Reviewer: ```suggestion
Alternatively, you can look SPNs up in the Attribute Editor of the Active Directory
```

> Users and Computers dialog on your AD joined Windows machine. RDS SQL Server

Reviewer: ```suggestion
Users and Computers dialog on your AD-joined Windows machine. The RDS SQL Server
```

logins that will use Active Directory authentication:

```sql
master> CREATE LOGIN [EXAMPLE\alice] FROM WINDOWS WITH DEFAULT_DATABASE = [master], DEFAULT_LANGUAGE = [us_english];
```

## Step 7/7. Connect

Log into your Teleport cluster. Your SQL Server database should appear in the
list of available databases:

```code
$ tsh login --proxy=teleport.example.com --user=alice
$ tsh db ls
# Name      Description         Labels
# --------- ------------------- -------
# sqlserver                     env=dev
```

Fetch short-lived client certificate for it using `tsh db login` command:

```code
$ tsh db login --db-user=teleport sqlserver
```

<Admonition type="tip">
  You can be logged into multiple databases simultaneously.
</Admonition>

Now connect to the database:

```code
$ tsh db connect sqlserver
```

<Admonition type="note">
  The `mssql-cli` command-line client should be available in PATH of the machine
  you're running `tsh db connect` from.
</Admonition>

To log out of the database and remove credentials:

```code
$ tsh db logout sqlserver
```

## Next steps

(!docs/pages/includes/database-access/guides-next-steps.mdx!)

## Further reading

> - [Manually join a Linux instance](https://docs.aws.amazon.com/directoryservice/latest/admin-guide/join_linux_instance.html) in AWS documentation.

Reviewer: ```suggestion
- [Manually join a Linux instance](https://docs.aws.amazon.com/directoryservice/latest/admin-guide/join_linux_instance.html) in the AWS documentation.
```

> - [Introduction to `adutil`](https://docs.microsoft.com/en-us/sql/linux/sql-server-linux-ad-auth-adutil-introduction) in Microsoft documentation.

Reviewer: ```suggestion
- [Introduction to `adutil`](https://docs.microsoft.com/en-us/sql/linux/sql-server-linux-ad-auth-adutil-introduction) in the Microsoft documentation.
```

## Step 1/7. Set up Teleport Auth and Proxy

(!docs/pages/includes/database-access/start-auth-proxy.mdx!)

## Step 2/7. Create a Teleport user

(!docs/pages/includes/database-access/create-user.mdx!)

## Step 3/7. Join the Linux node to Active Directory

<Admonition type="note">
  You can skip this step if you already have a Linux node joined to the same
  Active Directory domain as your SQL Server.
</Admonition>

The Linux node where the Database Service will run must be joined to the same
Active Directory domain as the SQL Server database.

Note that in order to be able to join, the Linux node must be able to resolve
your Active Directory fully-qualified domain name. For example, for AWS-managed
AD, use nameservers provided under "Networking details" on the directory's
overview page.

Install necessary packages:

<Tabs>
  <TabItem label="Ubuntu">
    ```bash
    $ sudo apt-get update
    $ sudo apt-get -y install sssd realmd krb5-user samba-common packagekit adcli
    ```
  </TabItem>
  <TabItem label="RHEL / CentOS 7">
    ```bash
    $ sudo yum -y update
    $ sudo yum -y install sssd realmd krb5-workstation samba-common-tools
    ```
  </TabItem>
</Tabs>

> Edit `/etc/krb5.conf` to disable reverse DNS resolution and set the default

Reviewer 2: Can we explicitly underline that  [realms] section should be also updated to point to default real ? 
```
[libdefaults]
 rdns = false
 default_realm = CROP.EXAMPLE.COM

[realms]
 CORP.EXAMPLE.COM = {
  kdc = corp.example.com
  admin_server = corp.example.com
 }
 ```

    --token=/tmp/token \
    --auth-server=teleport.example.com:3080 \
    --name=sqlserver \
    --protocol=sqlserver \
    --uri=sqlserver.example.com:1433 \
    --ad-keytab-file=/path/to/teleport.keytab \
    --ad-domain=EXAMPLE.COM \
    --ad-spn=MSSQLSvc/sqlserver.example.com:1433 \
    --labels=env=dev
  ```

Provide Active Directory parameters:

| Flag | Description |
| ---- | ----------- |
| `--ad-keytab-file` | Path to Kerberos keytab file generated above. |
| `--ad-domain` | Active Directory domain (Kerberos realm) that SQL Server is joined. |
| `--ad-spn` | Service Principal Name for SQL Server to fetch Kerberos tickets for. |

<Admonition type="tip">
  You can start Teleport Database Service using a configuration file instead of
  CLI flags. See [YAML reference](../reference/configuration.mdx).
</Admonition>

### Service Principal Name

You can use `adutil spn search` command to see the SPNs registered for your SQL
Server. Typically, they take a form of `MSSQLSvc/<name>.<ad-domain>:<port>`.

For example, an AWS RDS SQL Server named `sqlserver` and joined to an AWS managed
Active Directory domain `EXAMPLE.COM` will have the following SPNs registered:

```bash

> $ adutil spn search

Reviewer 2: I have found `adutil spn search` command unreliable but was able to query spn using ldapserach from `openldap-clients`package: 

```
ldapsearch  -x -h crop.example.com  -D test -W -b DC=crop,DC=example,DC=com  '(&(servicePrincipalName=*)(|(&(objectClass=user)(objectCategor^Cperson))(objectClass=group)(objectClass=computer)))'
```

...
CN=EC2AMAZ-4KN05DU,OU=RDS,OU=AWS Reserved,DC=example,DC=com:
    MSSQLSvc/sqlserver.example.com:1433
    MSSQLSvc/EC2AMAZ-4KN05DU.example.com:1433
    MSSQLSvc/EC2AMAZ-4KN05DU.example.com
...
```

Alternatively, you can look SPNs up in the Attribute Editor of the Active Directory
Users and Computers dialog on your AD-joined Windows machine. The RDS SQL Server
object typically resides under the AWS Reserved / RDS path:

> ![SPN](../../../img/database-access/guides/sqlserver/spn@2x.png)

Reviewer 2: The "Attribute Editor" tab is present only when "View>Advanced Features" is enabled.

You will be prompted to enter each user's password. All keytab entries will
be merged into the same `teleport.keytab` file.

<Admonition type="note" title="Assign Service Principal Names">
  For the `adutil keytab create` command to work, each user account must be
  assigned a Service Principal Name, otherwise the command will not be able
  to determine its `kvno` (key version number).

>   To check if the user has any SPNs assigned by running the following command

Reviewer: ```suggestion
To check if the user has any SPNs assigned, run the following command
```

# sqlserver                     env=dev
```

> Fetch the short-lived client certificate for it using `tsh db login` command:

Reviewer: ```suggestion
Fetch the short-lived client certificate for it using the `tsh db login` command:
```
